nemo/collections/tts/modules/transformer_2412.py

```diff
@@ -0,0 +1,704 @@
+            max_length_causal_mask <int>: Maximum length of causal mask
+            conv_non_linearity <Callable>: Convolution non-linearity
+        """
+        super().__init__()
+        self.has_xattn = has_xattn
+
+        self.norm_self = torch.nn.LayerNorm(d_model, bias=False)
+        self.self_attention = SelfAttention(
+            n_heads=sa_n_heads,
+            d_model=d_model,
+            p_dropout=p_dropout,
+            max_length_causal_mask=max_length_causal_mask,
+            is_causal=is_causal,
+        )
+
+        if self.has_xattn:
+            self.apply_norm_to_cond = apply_norm_to_cond
+            self.norm_xattn_query = torch.nn.LayerNorm(d_model, bias=False)
+            self.cross_attention = CrossAttention(
+                n_heads=xa_n_heads,
+                d_model=d_model,
+                d_memory=xa_d_memory,
+                p_dropout=p_dropout,
+            )
+
+            if self.apply_norm_to_cond:
+                self.norm_xattn_memory = torch.nn.LayerNorm(xa_d_memory, bias=False)
+
+        self.norm_pos_ff = torch.nn.LayerNorm(d_model, bias=False)
+        self.pos_ff = PositionwiseConvFF(
+            d_model, d_ffn, p_dropout, kernel_size=kernel_size, is_causal=is_causal, non_linearity=conv_non_linearity
+        )
+
+        self.use_cache = False
+        self.cache = self._init_cache()
+
+    @staticmethod
+    def _init_cache() -> Dict:
+        return {
+            'self_attn_output': None,
+            'cross_attn_output': None,
+            'memory': None,
+        }
+
+    def reset_cache(self, use_cache=False):
+        self.use_cache = use_cache
+        self.cache = self._init_cache()
+        self.self_attention.reset_cache(use_cache)
+        if self.has_xattn:
+            self.cross_attention.reset_cache(use_cache)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        x_mask: torch.Tensor,
+        cond: Optional[torch.Tensor] = None,
+        cond_mask: Optional[torch.Tensor] = None,
+        attn_prior: Optional[torch.Tensor] = None,
+    ) -> Dict:
+        """
+        Args:
+            x <torch tensor> (B, T1, C): Input tensor
+            x_mask <bool mask> (B, T1): True where ignoring is required
+            cond <torch tensor> (B, T2, C): Conditioning tensor
+            cond_mask <bool mask> (B, T2): True where ignoring is required
+
+        Returns dict with keys
+            output <torch tensor> (B, T1, C): Output tensor
+            attn_probabilities <dict>: Attention probabilities
+        """
+        x_mask_inv_float = (~x_mask).to(x.dtype).unsqueeze(-1)
+        s_attn_prob = None
+        x_, s_attn_prob = self.self_attention(query=self.norm_self(x), query_mask=x_mask)
+        if self.use_cache:
+            if self.cache['self_attn_output'] is not None:
+                x_ = torch.cat([self.cache['self_attn_output'], x_], dim=1)
+            self.cache['self_attn_output'] = x_
+        x = x + x_
+        x = x * x_mask_inv_float
```

Comment:
better to combine the addition and multiplication ops into a single one. it can reduce the number of separate tensor ops, minimizing the number of kernel launches. Single op creates fewer intermediate tensors, potentially reducing mem usage (see https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html#fuse-operations).
```suggestion
        x = (x + x_) * x_mask_inv_float
```